flash-patcher.py

```diff
@@ -220,293 +351,450 @@ def apply_patch(patch_file):
+            iface.recompile_data(part, output, output)
+
+
+def clean_scripts(modified_scripts: set) -> None:
+    """Delete all non-modified scripts.
+
+    Taken from https://stackoverflow.com/questions/19309667/recursive-os-listdir
+    - Make recursive os.listdir.
+    """
+    scripts = [
+        Path(dp, f) for dp, dn, fn in os.walk(DECOMP_LOCATION.expanduser()) for f in fn
+    ]
+    for script in scripts:
+        if script not in modified_scripts:
+            script.unlink()
 
-def main(inputfile, folder, stagefile, output, invalidate_cache, recompile_all, xml_mode):
-    print("Riley's SWF Patcher - " + CURRENT_VERSION)
 
-    if detect_jpexs() == False:
-        perror("Could not locate required dependency: JPEXS Flash Decompiler. Aborting...")
-        exit(1)
+def apply_patches(patches: list, folder: Path) -> set:
+    """Apply every patch, ignoring comments and empty lines."""
+    modified_scripts = set()
+    for patch in patches:
+        patch_stripped = patch.strip("\r\n ")
+        if len(patch_stripped) == 0 or patch_stripped[0] == "#":
+            continue
 
-    print("Using JPEXS at:", JPEXS_PATH)
+        # Check file extension of file
+        if patch_stripped.endswith(".patch"):  # Patch (code) file
+            modified_scripts |= apply_patch(folder / patch_stripped)
+        elif patch_stripped.endswith(".assets"):  # Asset Pack file
+            modified_scripts |= apply_assets(folder / patch_stripped, folder)
+        else:
+            exception(
+                "The file provided ('",
+                patch_stripped,
+                "') did not have a valid filetype.\n",
+                "Aborting...",
+            )
+            sys.exit(1)
 
-    cache_location = decompile_swf(inputfile, invalidate_cache, xml_mode)
+    return modified_scripts
+
+
+def main(
+    inputfile: Path,
+    folder: Path,
+    stagefile: Path,
+    output: Path,
+    *,
+    drop_cache: bool,
+    recompile_all: bool,
+    xml_mode: bool,
+) -> None:
+    """Run the patcher."""
+    info("Riley's SWF Patcher - %s", CURRENT_VERSION)
+
+    jpexs_iface = detect_jpexs()
+    if jpexs_iface is None:
+        error(
+            "Could not locate required dependency: JPEXS Flash Decompiler. Aborting...",
+        )
+        sys.exit(1)
+
+    info("Using JPEXS at: %s", jpexs_iface.path)
+
+    cache_location = decompile_swf(
+        jpexs_iface,
+        inputfile,
+        drop_cache=drop_cache,
+        xml_mode=xml_mode,
+    )
 
     # Copy the cache to a different location so we can reuse it
-    # os.path.isdir doesn't work on Linux for some reason, so we have to use a try-catch block
-    if (os.path.exists(DECOMP_LOCATION)):
-        try:
+    isdir = cache_location.is_dir()
+    if DECOMP_LOCATION.exists():
+        if isdir:
             shutil.rmtree(DECOMP_LOCATION)
-        except NotADirectoryError:
-            os.remove(DECOMP_LOCATION)
+        else:
+            Path.unlink(DECOMP_LOCATION)
```

Comment:
We should delete instead of unlinking here if possible, and warn the user that we're deleting `.Patcher-Temp`

Reply:
What's the difference between unlink and delete?

Reply:
unlink unlinks the inode, but leaves a floating file
some technical OS stuff: every file is an inode, and file names are just pointers to the inode. This allows for the creation of stuff like hardlinks (essentially shortcuts). When we run unlink instead of delete, we run the risk of creating an inode with no references. On some systems, this may cause the inode to be permanently allocated, but unreachable.
(Deleting a shortcut doesn't delete the inode, just the shortcut.)

See also: https://en.wikipedia.org/wiki/Inode

Reply:
I think the pathlib implementation does both, despite calling it `unlink`: https://docs.python.org/3/library/pathlib.html#pathlib.Path.unlink

Reply:
yep, we're good! just tested it locally